src/Lean/Compiler/IR/LLVMBindings.lean

```diff
@@ -0,0 +1,301 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddharth Bhat
+
+This file defines bindings for LLVM. The main mechanisms
+are to:
+- Mirror the values of C enumerations.
+- Create opaque values for LLVM's pointer based API
+- Write bindings that are `IO` based, which mutate the underlying in-memory
+  data structures to build IR.
+-/
+
+namespace LLVM
+
+namespace CodeGenFileType
+-- https://github.com/llvm/llvm-project/blob/c3e073bcbdc523b0f758d44a89a6333e38bff863/llvm/include/llvm-c/TargetMachine.h#L64
+def AssemblyFile : UInt64 := 0
+def ObjectFile : UInt64 := 1
+end CodeGenFileType
+
+namespace IntPredicate
+-- https://github.com/llvm/llvm-project/blob/c3e073bcbdc523b0f758d44a89a6333e38bff863/llvm/include/llvm-c/Core.h#L290
+def EQ : UInt64 := 32
+def NE : UInt64 := EQ + 1
+def UGT : UInt64 := NE + 1
+end IntPredicate
+
+inductive Ty where
+| int: Nat → Ty
+| float: Nat → Ty
+| ptr: Ty → Ty
+
+def Ty.i32: Ty := Ty.int 32
+def Ty.i1: Ty := Ty.int 1
+def Ty.i8: Ty := Ty.int 8
+def Ty.i8ptr: Ty := .ptr <| Ty.i8
+
+structure BasicBlock where
+structure Context where
+structure Module where
+structure Builder where
+structure LLVMType where
+structure Value where
+structure MemoryBuffer where
+structure Target where
+structure TargetMachine where
+
+-- A raw pointer to a C object, whose Lean representation
+-- is given by α
+opaque Ptr (α: Type): Type := Unit
+
+@[extern "lean_llvm_create_context"]
+opaque createContext: IO (Ptr Context)
+
+@[extern "lean_llvm_create_module"]
+opaque createModule (ctx: @&Ptr Context) (name: @&String): IO (Ptr Module)
+
+@[extern "lean_llvm_module_to_string"]
+opaque moduleToString (m: @&Ptr Module): IO String
+
+@[extern "lean_llvm_write_bitcode_to_file"]
+opaque writeBitcodeToFile (m: @&Ptr Module) (path: @&String): IO Unit
+
+@[extern "lean_llvm_add_function"]
+opaque addFunction (m: @&Ptr Module) (name: @&String) (type: @&Ptr LLVMType): IO (Ptr Value)
+
+@[extern "lean_llvm_get_named_function"]
+opaque getNamedFunction (m: @&Ptr Module) (name: @&String): IO (Option (Ptr Value))
+
+@[extern "lean_llvm_add_global"]
+opaque addGlobal (m: @&Ptr Module) (name: @&String) (type: @&Ptr LLVMType): IO (Ptr Value)
+
+@[extern "lean_llvm_get_named_global"]
+opaque getNamedGlobal (m: @&Ptr Module) (name: @&String): IO (Option (Ptr Value))
+
+@[extern "lean_llvm_build_global_string"]
+opaque buildGlobalString (builder: @&Ptr Builder) (value: @&String) (name: @&String := ""): IO (Ptr Value)
+
+@[extern "lean_llvm_set_initializer"]
+opaque setInitializer (glbl: @&Ptr Value) (val: @&Ptr Value): IO Unit
+
+@[extern "lean_llvm_function_type"]
+opaque functionType (retty: @&Ptr LLVMType) (args: @&Array (Ptr LLVMType)) (isVarArg: Bool := false): IO (Ptr LLVMType)
+
+@[extern "lean_llvm_void_type_in_context"]
+opaque voidType (ctx: @&Ptr Context): IO (Ptr LLVMType)
+
+@[extern "lean_llvm_int_type_in_context"]
+opaque intTypeInContext (ctx: @&Ptr Context) (width: UInt64): IO (Ptr LLVMType)
```

Comment:
Could functions like this be considered pure? I assume that would simplify the code a little. Could also be a function in namespace `Context`.